### ISSUE TITLE
Add MultiScheduler

### ISSUE THREAD

**Issue by 0phoff** [pytorch/pytorch#50282]
## 🚀 Feature
I would like to have a MultiScheduler that allows me to use multiple schedulers after each other, with the same optimizer.

## Motivation
It is not uncommon to have a learning rate schedule that looks like the following example:
- First have a burn in learning rate for 1000-2000 batches (eg. quadratic or exponential curve)
- Step down learning rate at steps 10.000 and 25.000.

You could implement all of this in a LambdaLR, but it would be a lot easier to combine a ``ExponentialLR`` and ``MultiStepLR``, running one after the other.

## Pitch
Implement a new LR-Scheduler, which I would call ``MultiScheduler`` (If anyone has a better name, I'm open to suggestions).
It receives a List of **N** schedulers, as well as a list **N-1** epoch numbers.
The scheduler can then automatically switch between different schedulers, depending on the `self.last_epoch` number.

I actually implemented such a scheduler in my own library and used it for quite a while now (ever since v0.4 IIRC), but I feel it deserves its place in the main PyTorch code base and am thus willing to make a PR that adds it!
However, unlike the current implementation, which takes a list of ``(epoch, scheduler)`` tuples, I think it would be cleaner take 2 separate lists.

## Alternatives
I do not think there are any viable alternatives to this problem, except for coding everything in a `LambdaLR` scheduler, which is far from ideal.

## Additional context
I don't know if there is any code in the PyTorch codebase that requires schedulers to have a ``self.optimizer`` variable, but as this scheduler combines multiple schedulers it has not.
However, I do use the base `_LRScheduler` class as a base and thus implement methods such as ``self.get_last_lr()``.

Let me know if you guys think it is a worthy addition to PyTorch, and I'll get started on my PR ! :construction_worker: 
Kind Regards,
0phoff

**Comment by 0phoff** [comment]
I created a draft PR, which can help to more clearly show what I would be implementing!

**Comment by vadimkantorov** [comment]
I also hit this. Also I needed a MultiOptimizer, since in the cases where we need different schedules for different part of the model, we also often need different optimizer (e.g. a simpler SGD for some parts). I implemented a wrapper that can be used for both goals, since scheduler and optimizer share API:

```python
class SteppableDict(dict):
    def step(self):
        for v in self.values();
            v.step()

    def state_dict(self):
        return {k : v.state_dict() for k, v in self.items()}

    def load_state_dict(self, state_dict, **kwargs):
        for k, v in self.items():
            v.load_state_dict(state_dict[k], **kwargs)
```

**Comment by danra** [comment]
For reference this is apparently ~[ChainedScheduler](https://docs.pytorch.org/docs/2.12/generated/torch.optim.lr_scheduler.ChainedScheduler.html)~ [SequentialLR](https://docs.pytorch.org/docs/2.12/generated/torch.optim.lr_scheduler.SequentialLR.html), available since v1.10 

Seems this can be closed as long done

**Comment by 0phoff** [comment]
> For reference this is apparently [ChainedScheduler](https://docs.pytorch.org/docs/2.12/generated/torch.optim.lr_scheduler.ChainedScheduler.html), available since v1.10
> 
> Seems this can be closed as long done

The ChainedScheduler runs `step()` on all schedulers each time, while my proposal was to have a scheduler that runs scheduler 1 for N epochs and then swap to scheduler 2, etc. Not entirely the same thing.

I am not in the AI research field anymore and won't have a direct need for it though, so fine for me to close with a "won't fix" but it is not implemented with the aforementioned ChainedScheduler as far as I understand it.

**Comment by danra** [comment]
Thanks @0phoff , fixed above

**Comment by 0phoff** [comment]
Ah yes that one indeed seems to be what I originally was after 😇
